### PR TITLE
IGNITE-13970 Thin client compatibility test

### PR DESCRIPTION
**Pull request by eadha**
This PR allows to run ducktape tests for thin client compatibility.

Server and client versions are provided as test parameters, default are dev + latest.
Test runs each client version on each server version and cheks the ordinary cache operations. 

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Comment by eadha** [comment]
Not correct target

File: modules/ducktests/tests/ignitetest/tests/thin_client_test.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+from ignitetest.services.ignite import IgniteService
+from ignitetest.services.ignite_app import IgniteApplicationService
+from ignitetest.services.utils.ignite_configuration import IgniteConfiguration
+from ignitetest.utils import ignite_versions, cluster
+from ignitetest.utils.ignite_test import IgniteTest
+from ignitetest.utils.version import DEV_BRANCH, LATEST, IgniteVersion
+
+class ThinClientTest(IgniteTest):
+    """
+    cluster - cluster size
```
Comment (nizhikov) [inline modules/ducktests/tests/ignitetest/tests/thin_client_test.py:32]: wong comments.

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/thin_client_test/ThinClientSelfTest.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import com.fasterxml.jackson.databind.JsonNode;
+import org.apache.ignite.cache.CacheAtomicityMode;
+import org.apache.ignite.cache.CacheMode;
+import org.apache.ignite.client.ClientCache;
+import org.apache.ignite.client.ClientCacheConfiguration;
+import org.apache.ignite.internal.ducktest.utils.IgniteAwareApplication;
+
+/**
+ * Thin client. Cache test: put, get, value check
```
Comment (nizhikov) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/thin_client_test/ThinClientSelfTest.java:28]: dot in the end.

File: modules/ducktests/tests/ignitetest/tests/thin_client_test.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    JAVA_CLIENT_CLASS_NAME = "org.apache.ignite.internal.ducktest.tests.thin_client_test.ThinClientSelfTest"
+
+    """
+    Tests services implementations
+    """
+
+    @cluster(num_nodes=2)
+    @ignite_versions(str(DEV_BRANCH))
+    @parametrize(num_nodes=2, clients_num=1)
+    def test_replicated_atomic_cache(self, ignite_version, num_nodes, clients_num):
```
Comment (nizhikov) [inline modules/ducktests/tests/ignitetest/tests/thin_client_test.py:47]: wrong test name.

Comment (nizhikov) [reply modules/ducktests/tests/ignitetest/tests/thin_client_test.py:47]: Let's remove `num_nodes` and `clients_num` and work with 2 nodes (1 server and 1 client) always.

File: modules/ducktests/tests/ignitetest/tests/thin_client_test.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+        static_clients = IgniteApplicationService(self.test_context, server_configuration,
+                                           java_class_name=self.JAVA_CLIENT_CLASS_NAME,
+                                           num_nodes=clients_num,
+                                           params={"thin_client_connection": thin_client_connection},
+                                           start_ignite = False,
+                                           startup_timeout_sec=180)
+
+        ignite.start()
+        static_clients.start()
```
Comment (nizhikov) [inline modules/ducktests/tests/ignitetest/tests/thin_client_test.py:67]: please, add await_started and await_finished checks here.
You should use the following methods - https://github.com/apache/ignite/blob/8c432ba136b5bdb4e6c7e4b3265785b53959d036/modules/ducktests/tests/ignitetest/services/ignite_app.py#L53

File: modules/ducktests/tests/ignitetest/services/ignite_app.py [hunk 5]
```diff
@@ -46,14 +46,17 @@ def __init__(self, context, config, java_class_name, num_nodes=1, params="", sta
         self.servicejava_class_name = servicejava_class_name
         self.java_class_name = java_class_name
         self.params = params
+        self.start_ignite = start_ignite
```
Comment (nizhikov) [inline modules/ducktests/tests/ignitetest/services/ignite_app.py:49]: Why this change?

File: modules/ducktests/tests/ignitetest/tests/thin_client_test.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    def test_thin_client_compatibility(self, server_version, thin_client_version):
+        """
+        Thin client compatibility test.
+        """
+
+        ignite = IgniteService(self.test_context, IgniteConfiguration(version=IgniteVersion(server_version), caches=[]), 1)
+
+        client_config = IgniteConfiguration(version=IgniteVersion(thin_client_version), caches=[])
+
+        thin_client_connection = ignite.nodes[0].account.hostname + ":" + str(client_config.client_connector_configuration.port)
```
Comment (nizhikov) [inline modules/ducktests/tests/ignitetest/tests/thin_client_test.py:49]: We should get port from *server* config.

**Review by nizhikov — APPROVED** [review]

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/utils/IgniteAwareApplicationService.java [hunk 7]
```diff
@@ -60,6 +61,11 @@ public static void main(String[] args) throws Exception {
 
         app.cfgPath = cfgPath;
 
+        String connStr = jsonNode.get("thin_client_connection").asText();
+
+        if (connStr != null && !connStr.isEmpty())
```
Comment (anton-vinogradov) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/utils/IgniteAwareApplicationService.java:66]: We should have a proper switch here.
`startIgnite` should be replaced with Enum [Node, ThinClient, None]
The second param in this case should contain configuration for ThinClient if the `ThinClient` case selected.

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/utils/IgniteAwareApplicationService.java [hunk 8]
```diff
@@ -60,6 +61,11 @@ public static void main(String[] args) throws Exception {
 
         app.cfgPath = cfgPath;
 
+        String connStr = jsonNode.get("thin_client_connection").asText();
```
Comment (anton-vinogradov) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/utils/IgniteAwareApplicationService.java:64]: Should be a constant at least. 
But, see next comment why this not needed.

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/thin_client_test/ThinClientSelfTest.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+        cache.put(0, 0);
+
+        Integer val = cache.get(0);
+
+        cache.remove(0);
+
+        client.destroyCache("testCache");
+
+        markFinished();
```
Comment (anton-vinogradov) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/thin_client_test/ThinClientSelfTest.java:42]: We should have at least one assertion here before finishing, eg. put == read.
Otherwise, Volkswagen mode is possible.

File: modules/ducktests/tests/ignitetest/services/utils/ignite_configuration/__init__.py [hunk 10]
```diff
@@ -45,6 +46,7 @@ class IgniteConfiguration(NamedTuple):
     local_host: str = None
     ssl_params: SslParams = None
     connector_configuration: ConnectorConfiguration = None
+    client_connector_configuration: ClientConnectorConfiguration = ClientConnectorConfiguration()
```
Comment (anton-vinogradov) [inline modules/ducktests/tests/ignitetest/services/utils/ignite_configuration/__init__.py:49]: Should this be enabled by explicit enabling?

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/utils/IgniteAwareApplicationService.java (hunk 8 above)
Comment (Sega76) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/utils/IgniteAwareApplicationService.java:64]: use Optional, otherwise we will get an Exception in other tests
`Exception in thread "main" java.lang.NullPointerException
	at org.apache.ignite.internal.ducktest.utils.IgniteAwareApplicationService.main(IgniteAwareApplicationService.java:64)`

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/thin_client_test/ThinClientSelfTest.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+public class ThinClientSelfTest extends IgniteAwareApplication {
+    /** {@inheritDoc} */
+    @Override protected void run(JsonNode jsonNode) throws Exception {
+        markInitialized();
+
+        ClientCache<Integer, Integer> cache = client.getOrCreateCache("testCache");
+
+        cache.put(0, 0);
+
+        Integer val = cache.get(0);
```
Comment (Sega76) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/tests/thin_client_test/ThinClientSelfTest.java:36]: unused `Integer val`

File: modules/ducktests/tests/ignitetest/services/ignite_app.py (hunk 5 above)
Comment (anton-vinogradov) [inline modules/ducktests/tests/ignitetest/services/ignite_app.py:49]: An incorrect place to store this flag.
The flag is never used inside the class.
Should be stored at IgniteAwareService.

File: modules/ducktests/tests/ignitetest/services/ignite_app.py [hunk 12]
```diff
@@ -46,14 +46,17 @@ def __init__(self, context, config, java_class_name, num_nodes=1, params="", sta
...
     def await_started(self):
-        super().await_started()
+        if self.start_ignite:
+            super().await_started()
 
         self.__check_status(self.APP_INIT_EVT_MSG, timeout=self.startup_timeout_sec)
 
     def await_stopped(self):
-        super().await_stopped()
+        if self.start_ignite:
```
Comment (anton-vinogradov) [inline modules/ducktests/tests/ignitetest/services/ignite_app.py:58]: Incorrect usage.
We still MUST wait for the Application to finish.

File: modules/ducktests/tests/ignitetest/tests/thin_client_test.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    def test_thin_client_compatibility(self, server_version, thin_client_version):
+        """
+        Thin client compatibility test.
+        """
+
+        server_config = IgniteConfiguration(version=IgniteVersion(server_version))
+
+        ignite = IgniteService(self.test_context, server_config, 1)
+
+        thin_client_connection = ignite.nodes[0].account.hostname + ":" + str(server_config.client_connector_configuration.port)
```
Comment (anton-vinogradov) [inline modules/ducktests/tests/ignitetest/tests/thin_client_test.py:49]: Use ThinClient config instead.

**Review by anton-vinogradov — CHANGES_REQUESTED** [review]

File: modules/ducktests/tests/ignitetest/services/ignite_app.py (hunk 5 above)
Comment (eadha) [reply modules/ducktests/tests/ignitetest/services/ignite_app.py:49]: it is used in await_started, string 52

Comment (anton-vinogradov) [reply modules/ducktests/tests/ignitetest/services/ignite_app.py:49]: As I said, de-facto, this usage can and *should* be encapsulated at `IgniteAwareService`.`await_started()`

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/utils/IgniteAwareApplicationService.java (hunk 7 above)
Comment (anton-vinogradov) [reply modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/utils/IgniteAwareApplicationService.java]: Additional reflection suggests we may want to test the local Thin Client case, so, enum is not an option here.
But, since `IgniteClient extends AutoCloseable` we must wrap the current try-with-resources section with an additional one.

File: modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py [hunk 14]
```diff
@@ -74,6 +74,7 @@ def __init__(self, context, config, num_nodes, startup_timeout_sec, shutdown_tim
         self.init_logs_attribute()
 
         self.disconnected_nodes = []
+        self.start_ignite = kwargs.get("start_ignite")
```
Comment (nizhikov) [inline modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py:77]: you should provide the default value for `start_ignite` - `kwargs.get("start_ignite", True)`

File: modules/ducktests/tests/ignitetest/services/ignite_app.py [hunk 15]
```diff
@@ -47,8 +47,9 @@ def __init__(self, context, config, java_class_name, num_nodes=1, params="", sta
         self.java_class_name = java_class_name
         self.params = params
 
-    def await_started(self):
-        super().await_started()
+    def await_started(self, start_ignite=True):
+        if start_ignite:
```
Comment (nizhikov) [inline modules/ducktests/tests/ignitetest/services/ignite_app.py:51]: please, move this check to `ignite_aware.py`

**Review by nizhikov — APPROVED** [review]

File: modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/utils/IgniteAwareApplicationService.java [hunk 16]
```diff
@@ -76,7 +81,18 @@ public static void main(String[] args) throws Exception {
...
+
+            try (IgniteClient client = Ignition.startClient(new ClientConfiguration().setAddresses(connStr))) {
+                app.client = client;
+
+                app.start(jsonNode);
+            }
+            finally {
+                log.info("Thin client instance closed. [interrupted=" + Thread.currentThread().isInterrupted() + "]");
+            }
+
```
Comment (nizhikov) [inline modules/ducktests/src/main/java/org/apache/ignite/internal/ducktest/utils/IgniteAwareApplicationService.java:95]: please, remove empty line.

File: modules/ducktests/tests/ignitetest/tests/thin_client_test.py [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+                                                java_class_name=self.JAVA_CLIENT_CLASS_NAME,
+                                                num_nodes=1,
+                                                params={"thin_client_connection": thin_client_connection},
+                                                start_ignite=False)
+
+        ignite.start()
+        thin_clients.run()
+
+        thin_clients.await_event(evt_message="IGNITE_APPLICATION_FINISHED", timeout_sec=60, from_the_beginning=True)
+        thin_clients.stop()
```
Comment (anton-vinogradov) [inline modules/ducktests/tests/ignitetest/tests/thin_client_test.py:64]: All we need is just to call service.run()
It will start, wait and stop the service.
Also, it will checks statuses for each state (eg. IGNITE_APPLICATION_FINISHED will be checked on stop phase)

**Review by anton-vinogradov — APPROVED** [review]
